// ink-multi-select/build/mutli-select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Indicator", {
  enumerable: true,
  get: function () {
    return _indicator.default;
  }
});
Object.defineProperty(exports, "Item", {
  enumerable: true,
  get: function () {
    return _item.default;
  }
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = _interopRequireDefault(require("lodash.isequal"));

var _arrRotate = _interopRequireDefault(require("arr-rotate"));

var _ink = require("ink");

var _indicator = _interopRequireDefault(require("./indicator"));

var _item = _interopRequireDefault(require("./item"));

var _checkbox = _interopRequireDefault(require("./checkbox"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const ARROW_UP = '\u001B[A';
const ARROW_DOWN = '\u001B[B';
const ENTER = '\r';
const SPACE = ' ';

class MultiSelect extends _react.PureComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      rotateIndex: 0,
      highlightedIndex: this.props.initialIndex,
      selected: {}
    });

    _defineProperty(this, "handleInput", data => {
      const {
        items,
        focus,
        onSelect,
        onUnselect,
        onHighlight,
        onSubmit
      } = this.props;
      const {
        rotateIndex,
        highlightedIndex,
        selected
      } = this.state;
      const {
        limit,
        hasLimit
      } = this;

      if (focus === false) {
        return;
      }

      const s = String(data);

      if (s === ARROW_UP || s === 'k') {
        const lastIndex = (hasLimit ? limit : items.length) - 1;
        const atFirstIndex = highlightedIndex === 0;
        const nextIndex = hasLimit ? highlightedIndex : lastIndex;
        const nextRotateIndex = atFirstIndex ? rotateIndex + 1 : rotateIndex;
        const nextHighlightedIndex = atFirstIndex ? nextIndex : highlightedIndex - 1;
        this.setState({
          rotateIndex: nextRotateIndex,
          highlightedIndex: nextHighlightedIndex
        });
        const slicedItems = hasLimit ? (0, _arrRotate.default)(items, nextRotateIndex).slice(0, limit) : items;
        onHighlight(slicedItems[nextHighlightedIndex]);
      }

      if (s === ARROW_DOWN || s === 'j') {
        const atLastIndex = highlightedIndex === (hasLimit ? limit : items.length) - 1;
        const nextIndex = hasLimit ? highlightedIndex : 0;
        const nextRotateIndex = atLastIndex ? rotateIndex - 1 : rotateIndex;
        const nextHighlightedIndex = atLastIndex ? nextIndex : highlightedIndex + 1;
        this.setState({
          rotateIndex: nextRotateIndex,
          highlightedIndex: nextHighlightedIndex
        });
        const slicedItems = hasLimit ? (0, _arrRotate.default)(items, nextRotateIndex).slice(0, limit) : items;
        onHighlight(slicedItems[nextHighlightedIndex]);
      }

      if (s === SPACE) {
        const slicedItems = hasLimit ? (0, _arrRotate.default)(items, rotateIndex).slice(0, limit) : items;
        const selectedItem = slicedItems[highlightedIndex];
        const selectedItemKey = selectedItem.key || selectedItem.value;
        (selected[selectedItemKey] ? onUnselect : onSelect)(selectedItem);
        this.setState(state => ({
          selected: { ...selected,
            [selectedItemKey]: !state.selected[selectedItemKey]
          }
        }));
      }

      if (s === ENTER) {
        onSubmit(items.filter(item => selected[item.key || item.value]));
      }
    });
  }

  render() {
    const {
      items,
      indicatorComponent,
      itemComponent
    } = this.props;
    const {
      rotateIndex,
      highlightedIndex,
      selected
    } = this.state;
    const {
      limit,
      hasLimit
    } = this;
    const slicedItems = hasLimit ? (0, _arrRotate.default)(items, rotateIndex).slice(0, limit) : items;
    return _react.default.createElement(_ink.Box, {
      flexDirection: "column"
    }, slicedItems.map((item, index) => {
      const key = item.key || item.value;
      const isHighlighted = index === highlightedIndex;
      const isSelected = Boolean(selected[key]);
      return _react.default.createElement(_ink.Box, {
        key: key
      }, _react.default.createElement(indicatorComponent, {
        isHighlighted
      }), _react.default.createElement(checkBoxComponent, {
        isSelected
      }), _react.default.createElement(itemComponent, { ...item,
        isHighlighted
      }));
    }));
  }

  componentDidMount() {
    const {
      stdin,
      setRawMode
    } = this.props; // eslint-disable-line react/prop-types

    setRawMode(true);
    stdin.on('data', this.handleInput);
  }

  componentWillUnmount() {
    const {
      stdin,
      setRawMode
    } = this.props; // eslint-disable-line react/prop-types

    stdin.removeListener('data', this.handleInput);
    setRawMode(false);
  }

  componentDidUpdate(prevProps) {
    if (!(0, _lodash.default)(prevProps.items, this.props.items)) {
      this.setState({
        // eslint-disable-line react/no-did-update-set-state
        rotateIndex: 0,
        highlightedIndex: 0
      });
    }
  }

  get hasLimit() {
    const {
      limit,
      items
    } = this.props;
    return typeof limit === 'number' && items.length > limit;
  }

  get limit() {
    const {
      limit,
      items
    } = this.props;

    if (this.hasLimit) {
      return Math.min(limit, items.length);
    }

    return items.length;
  }

}

_defineProperty(MultiSelect, "propTypes", {
  items: _propTypes.default.array,
  focus: _propTypes.default.bool,
  initialIndex: _propTypes.default.number,
  indicatorComponent: _propTypes.default.func,
  checkBoxComponent: _propTypes.default.func,
  itemComponent: _propTypes.default.func,
  limit: _propTypes.default.number,
  onSelect: _propTypes.default.func,
  onUnselect: _propTypes.default.func,
  onSubmit: _propTypes.default.func,
  onHighlight: _propTypes.default.func
});

_defineProperty(MultiSelect, "defaultProps", {
  items: [],
  focus: true,
  initialIndex: 0,
  indicatorComponent: _indicator.default,
  checkBoxComponent: _checkbox.default,
  itemComponent: _item.default,
  limit: null,

  onSelect() {},

  onUnselect() {},

  onSubmit() {},

  onHighlight() {}

});

class MultiSelectWithStdin extends _react.PureComponent {
  render() {
    return _react.default.createElement(_ink.StdinContext.Consumer, null, ({
      stdin,
      setRawMode
    }) => _react.default.createElement(MultiSelect, _extends({}, this.props, {
      stdin: stdin,
      setRawMode: setRawMode
    })));
  }

}

exports.default = MultiSelectWithStdin;
